<a href="https://colab.research.google.com/github/Inazuma1002/Web-Science-Lab---Jan2024-Jun2024/blob/main/Notebooks/forward.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sea
from sklearn.model_selection import train_test_split as tts
from sklearn.linear_model import LinearRegression

In [ ]:
data = pd.read_excel('/content/drive/MyDrive/ML/Excel sheets/Hospital_facilities_2018-22.xlsx');
print(data)

In [ ]:
for col in data.keys():
  if col not in {'Sr no', 'Districts'}:
    data[f'Delta_{col}'] = data[col].diff()
data['Delta_Maternal Death'] = data['Maternal Death'].diff()
print (data)

In [ ]:
plt.figure(figsize=(10, 10))

plt.scatter(data['Delta_108 Ambulances_No. of Patients benefited'], data['Delta_Maternal Death'], label='change in MD due to change in Number of patients benefited of Ambulance')
plt.scatter(data['Delta_Number of 24/7 working Hospitals_No. of Delivaries occured in Hospitals'], data['Delta_Maternal Death'], label = 'change in MD due to change in Deliveries happening in hospitals' )
plt.xlabel('Change in Attribute')
plt.ylabel('Change in Maternal Death rate')
plt.title('Change in Attribute vs. Change in Maternal Death over 4 years')
plt.legend()
plt.grid(True)
plt.show()

In [ ]:
df = pd.DataFrame(data)
df = df.dropna()
df

In [ ]:
X = df.drop(['Sr no', 'Districts', 'F.R.U\n(First Referral Units)', '108 Ambulances_No. of Vehicles', '108 Ambulances_No. of Patients benefited', 'Number of 24/7 working Hospitals_Nos', 'Number of 24/7 working Hospitals_No. of Delivaries occured in Hospitals', 'No. of Benificiaries availed Janani Suraksha Yojana', 'No. of Beneficiaries received Madilu kit', 'Maternal Death', 'No. of Beneficiaries under Supplementary Nutrition Programme','Delta_Maternal Death'], axis=1)  # Predictor variables
y = df['Delta_Maternal Death']  # Target variable

In [ ]:
model = LinearRegression()


In [ ]:
coefficients = dict(zip(X.columns, model.coef_))
print("Weightage of changes in each attribute on changes in the target variable over 4 years:")
print(coefficients)

In [ ]:
data2 = pd.read_excel('/content/drive/MyDrive/ML/Excel sheets/Districts_2018-22/Bagalkot.xlsx')
for col in data2.keys():
  if col not in {'Sr no', 'Districts'}:
    data2[f'Delta_{col}'] = data2[col].diff()
data2['Delta_Maternal Death'] = data2['Maternal Death'].diff()
print (data2)

In [ ]:
df = pd.DataFrame(data2)
df = df.dropna()
X = df.drop(['Sr no', 'Districts', 'F.R.U\n(First Referral Units)', '108 Ambulances_No. of Vehicles', '108 Ambulances_No. of Patients benefited', 'Number of 24/7 working Hospitals_Nos', 'Number of 24/7 working Hospitals_No. of Delivaries occured in Hospitals', 'No. of Benificiaries availed Janani Suraksha Yojana', 'No. of Beneficiaries received Madilu kit', 'Maternal Death', 'No. of Beneficiaries under Supplementary Nutrition Programme','Delta_Maternal Death'], axis=1)  # Predictor variables
y = df['Delta_Maternal Death']  # Target variable
model = LinearRegression()
model.fit(X, y)
coefficients2 = dict(zip(X.columns, model.coef_))
print("Weightage of changes in each attribute on changes in the target variable over 4 years:")
print(coefficients2)

In [ ]:
average_coefficients = {}

for key in coefficients.keys():
    average_coefficient = (coefficients[key] + coefficients2[key]) / 2
    print(f"Average coefficient for {key}: {average_coefficient}")

April 23

In [ ]:
data = pd.read_excel('/content/drive/MyDrive/ML/Excel sheets/Total Districts-2.xlsx')
data

In [ ]:
for col in data.keys():
  if col not in {'Sr no', 'Districts'}:
    data[f'Delta_{col}'] = data[col].diff()
data['Delta_Maternal Death'] = data['Maternal Death'].diff()
print (data)

In [ ]:
print(df.dtypes)

In [ ]:
df = pd.DataFrame(data)

# Min-max scaling for each attribute
for col in df.columns:
    if col not in {'Districts', 'Sr no'} :
        min_val = df[col].min()
        max_val = df[col].max()
        df[col + '_norm'] = (df[col] - min_val) / (max_val - min_val)
df

In [ ]:
for col in data.keys():
  if col not in {'Sr no', 'Districts','108 Ambulances_No. of Vehicles','108 Ambulances_No. of Patients benefite','Number of 24/7 working Hospitals_Nos','Number of 24/7 working Hospitals_No. of Delivaries occured in Hospitals','No. of Benificiaries availed Janani Suraksha Yojana','No. of Beneficiaries received Madilu kit','Maternal Death', 'No. of Beneficiaries under Supplementary Nutrition Programme'}:
    data[f'Delta_{col}'] = data[col].diff()
data['Delta_Maternal Death'] = data['Maternal Death'].diff()

data

In [ ]:
df = pd.DataFrame(data)
df = df.dropna()
X = df.drop(['Sr no', 'Districts', 'F.R.U\n(First Referral Units)', '108 Ambulances_No. of Vehicles', '108 Ambulances_No. of Patients benefited', 'Number of 24/7 working Hospitals_Nos', 'Number of 24/7 working Hospitals_No. of Delivaries occured in Hospitals', 'No. of Benificiaries availed Janani Suraksha Yojana', 'No. of Beneficiaries received Madilu kit', 'Maternal Death', 'No. of Beneficiaries under Supplementary Nutrition Programme','Delta_Maternal Death'], axis=1)  # Predictor variables
y = df['Delta_Maternal Death']  # Target variable
model = LinearRegression()
model.fit(X, y)
coefficients2 = dict(zip(X.columns, model.coef_))
coefficients2


May 1


In [ ]:
data = pd.read_excel('/content/drive/MyDrive/ML/Excel sheets/May 1.xlsx')
data.dropna(inplace = True)
data

In [ ]:
df = pd.DataFrame(data)
df = df.dropna()
X = df.drop(['Maternal Death'], axis=1)  # Predictor variables
y = df['Maternal Death']  # Target variable
model = LinearRegression()
model.fit(X, y)
coefficients2 = dict(zip(X.columns, model.coef_))
coefficients2

{'F.R.U\n(First Referral Units)': 0.41622232577856844,
 '108 Ambulances_No. of Vehicles': -0.13602850367970107,
 '108 Ambulances_No. of Patients benefited': -0.0007257719457827161,
 'Number of 24/7 working Hospitals_Nos': 0.04894077660013333,
 'Number of 24/7 working Hospitals_No. of Delivaries occured in Hospitals': 0.0001669299169968913,
 'No. of Benificiaries availed Janani Suraksha Yojana': -0.00036948518498678123,
 'No. of Beneficiaries received Madilu kit': -0.0017136945926175828,
 'No. of Beneficiaries under Supplementary Nutrition Programme': -0.00012768096293563295}

In [ ]:
import statsmodels.api as sm

In [ ]:
X = df.drop(['Maternal Death'], axis=1)  # Predictor variables
y = df['Maternal Death']  # Target variable
X = sm.add_constant(X)
model = sm.OLS(y, X).fit()

# Print regression results
print(model.summary())

                            OLS Regression Results                            
Dep. Variable:         Maternal Death   R-squared:                       0.637
Model:                            OLS   Adj. R-squared:                  0.499
Method:                 Least Squares   F-statistic:                     4.615
Date:                Wed, 01 May 2024   Prob (F-statistic):            0.00230
Time:                        10:50:20   Log-Likelihood:                -118.72
No. Observations:                  30   AIC:                             255.4
Df Residuals:                      21   BIC:                             268.1
Df Model:                           8                                         
Covariance Type:            nonrobust                                         
                                                                              coef    std err          t      P>|t|      [0.025      0.975]
----------------------------------------------------------------------

1. The coefficients represent the estimated change in the dependent variable for a one-unit change in the corresponding independent variable, holding other variables constant.

2. For example, the coefficient for "F.R.U (First Referral Units)" is 0.4162. This suggests that, on average, a one-unit increase in the number of First Referral Units is associated with a 0.4162 unit increase in Maternal Death.


3. The p-values associated with each coefficient indicate the statistical significance of the relationship between the independent variable and the dependent variable. For instance, the coefficient for "108 Ambulances_No. of Patients benefited" has a p-value of 0.000, suggesting a statistically significant relationship.

R-squared value of 0.637 means that approximately 63.7% of the variance in the dependent variable (Maternal Death) is explained by the independent variables included in the model.

According to P-value, In summary, only the variable "108 Ambulances_No. of Patients benefited" is statistically significant at the conventional significance level of 0.05. The number of patients benefited by ambulances is negatively associated with Maternal Death. Other variables do not have statistically significant relationships with Maternal Death based on the given results.

In [ ]:
import pdfplumber

# Open the PDF file
with pdfplumber.open("your_pdf_file.pdf") as pdf:
    # Iterate through each page
    for page in pdf.pages:
        # Extract tables from the page
        tables = page.extract_tables()

        # Iterate through each table on the page
        for table in tables:
            # Print the table
            for row in table:
                print(row)
            print("\n")  # Separate tables with a blank line
